### PR TITLE
state: make storage persistent by default

### PR DESCRIPTION
**Pull request by axw**
Volumes and filesystems will not default to
being bound to the model, or machine, depending
on whether or not the storage can be persistent.

This is the minimal change. Volume-backed
filesystems are currently always considered
machine-scoped, which leads to undesirable
behaviour: the filesystem cannot be detached
from the machine it is initially attached to
(or doing so will cause the volume to be
destroyed). We will need to make further
changes so that volume-backed filesystems
are not scoped or bound to a single machine.

### CODE REVIEW

**Comment by axw** [comment]
**QA**

(combined with https://github.com/juju/juju/pull/6673)

1. juju bootstrap google
2. juju deploy cs:~prometheus-charmers/prometheus --storage metrics-filesystem=gce,10G
3. try to remove volume related to prometheus storage, get an error saying that it contains a filesystem
4. try to remove filesystem related to prometheus storage, get an error saying that it has assigned storage
5. remove storage from prometheus (successful), leaves both volume and filesystem behind
6. remove filesystem (successful), also detaches/destroys/removes volume

**Comment by anastasiamac** [comment]
What are doing when we are removing units/application with persistent storage? Do we need to ask user confirmation whether persistent storage needs to be removed? Do we default to removing it unless there is an explicit "get rid of persistent storage" option set?

**Comment by axw** [comment]
> What are doing when we are removing units/application with persistent storage? Do we need to ask user confirmation whether persistent storage needs to be removed? Do we default to removing it unless there is an explicit "get rid of persistent storage" option set?

When this change goes ahead, we'll only remove the storage when (a) the user explicitly says so (runs a remove- command), or (b) when the model is destroyed (or machine, for inherently machine-bound storage, such as loop, tmpfs, etc.).

We may want to extend the remove-application/remove-unit/remove-machine commands later to give the user a chance to make a choice up front; that can be layered on top of this change.

File: state/cleanup.go [hunk 1]
```diff
@@ -212,7 +218,41 @@ func (st *State) cleanupMachinesForDyingModel() (err error) {
...
+	for _, fs := range filesystems {
+		err := st.DestroyFilesystem(fs.FilesystemTag())
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// cleanupApplicationsForDyingModel sets all services to Dying, if they are
```
Comment (wallyworld) [inline state/cleanup.go]: can we drive-by change services to application here?

File: state/filesystem.go [hunk 2]
```diff
@@ -588,13 +599,20 @@ func removeFilesystemAttachmentOps(m names.MachineTag, f *filesystem) []txn.Op {
...
 	buildTxn := func(attempt int) ([]txn.Op, error) {
 		filesystem, err := st.filesystemByTag(tag)
 		if errors.IsNotFound(err) {
 			return nil, jujutxn.ErrNoOperations
 		} else if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if filesystem.doc.StorageId != "" {
+			return nil, errors.Errorf(
+				"filesystem is assigned to %s",
```
Comment (wallyworld) [inline state/filesystem.go]: it would be useful to include the machine/unit in the error message as that's how users tend think of the scope of storage

File: state/cleanup.go [hunk 3]
```diff
@@ -212,7 +218,41 @@ func (st *State) cleanupMachinesForDyingModel() (err error) {
...
+// a model is destroyed.
+func (st *State) cleanupVolumesForDyingModel() (err error) {
+	volumes, err := st.AllVolumes()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, v := range volumes {
+		err := st.DestroyVolume(v.VolumeTag())
+		if err != nil && !IsContainsFilesystem(err) {
+			return errors.Trace(err)
```
Comment (wallyworld) [inline state/cleanup.go]: I wonder if this error should be typed so that we can avoid log noise. Wouldn't it be the case that the filesystem cleanup would take care of the underlying volume? So a dying model should see both filesystems and volumes destroyed as expected.

File: state/filesystem.go [hunk 4]
```diff
@@ -782,24 +809,34 @@ func (st *State) newFilesystemOps(doc filesystemDoc, status statusDoc) []txn.Op
...
-	if err != nil {
-		return FilesystemParams{}, errors.Trace(err)
-	}
-	cons := StorageConstraints{
-		Pool:  params.Pool,
-		Size:  params.Size,
-		Count: 1,
+func (st *State) filesystemParamsWithDefaults(params FilesystemParams, machineId string) (FilesystemParams, error) {
+	if params.Pool == "" {
+		envConfig, err := st.ModelConfig()
```
Comment (wallyworld) [inline state/filesystem.go]: nit: modelConfig

File: state/filesystem_test.go [hunk 5]
```diff
@@ -533,6 +533,24 @@ func (s *FilesystemStateSuite) TestDestroyFilesystem(c *gc.C) {
...
+	// cannot destroy the filesystem while there is storage assigned.
+	_, u, storageTag := s.setupSingleStorage(c, "filesystem", "rootfs")
+	err := s.State.AssignUnit(u, state.AssignCleanEmpty)
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = u.AssignedMachineId()
+	c.Assert(err, jc.ErrorIsNil)
+	filesystem := s.storageInstanceFilesystem(c, storageTag)
+
+	err = s.State.DestroyFilesystem(filesystem.FilesystemTag())
+	c.Assert(err, gc.ErrorMatches, "destroying filesystem 0/0: filesystem is assigned to storage data/0")
```
Comment (wallyworld) [inline state/filesystem_test.go]: As per earlier comment, including machine/unit would be helpful here for the user

File: state/filesystem_test.go [hunk 6]
```diff
@@ -533,6 +533,24 @@ func (s *FilesystemStateSuite) TestDestroyFilesystem(c *gc.C) {
...
+
+	removeStorageInstance(c, s.State, storageTag)
+	err = s.State.DestroyFilesystem(filesystem.FilesystemTag())
```
Comment (wallyworld) [inline state/filesystem_test.go]: can we replace these 2 lines with assertDestroy() to do a proper check

File: state/model.go [hunk 7]
```diff
@@ -95,6 +95,12 @@ type modelEntityRefsDoc struct {
 
 	// Applicatons contains the names of the applications in the model.
 	Applications []string `bson:"applications"`
+
+	// Volumes contains the IDs of the volumes in the model.
```
Comment (wallyworld) [inline state/model.go]: perhaps s/in the model/bound to the model
to be consistent with terminology used elsewhere?

File: state/model_test.go [hunk 8]
```diff
@@ -661,6 +661,81 @@ func (s *ModelSuite) TestProcessDyingEnvironWithMachinesAndServicesNoOp(c *gc.C)
 	c.Assert(env.Destroy(), jc.ErrorIsNil)
 }
 
+func (s *ModelSuite) TestProcessDyingEnvironWithVolumeBackedFilesystems(c *gc.C) {
```
Comment (wallyworld) [inline state/model_test.go]: DyingModel

File: state/model_test.go [hunk 9]
```diff
@@ -661,6 +661,81 @@ func (s *ModelSuite) TestProcessDyingEnvironWithMachinesAndServicesNoOp(c *gc.C)
...
+	c.Assert(volumes, gc.HasLen, 1)
+
+	c.Assert(model.Destroy(), jc.ErrorIsNil)
+	err = st.DetachVolume(machine.MachineTag(), names.NewVolumeTag("0"))
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.RemoveVolumeAttachment(machine.MachineTag(), names.NewVolumeTag("0"))
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(machine.EnsureDead(), jc.ErrorIsNil)
+	c.Assert(machine.Remove(), jc.ErrorIsNil)
+	err = st.ProcessDyingModel()
```
Comment (wallyworld) [inline state/model_test.go]: can we add a comment similar to the previous test about volume being persistent

Comment (wallyworld) [inline state/model_test.go]: doesn't there need to be tests for the non-persistent cases also?

File: state/volume.go [hunk 10]
```diff
@@ -655,6 +667,12 @@ func (st *State) DestroyVolume(tag names.VolumeTag) (err error) {
 		} else if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if volume.doc.StorageId != "" {
+			return nil, errors.Errorf(
+				"volume is assigned to %s",
```
Comment (wallyworld) [inline state/volume.go]: include machine/unit here as well

File: state/volume.go [hunk 11]
```diff
@@ -770,27 +791,38 @@ func (st *State) newVolumeOps(doc volumeDoc, status statusDoc) []txn.Op {
...
-	if err != nil {
-		return VolumeParams{}, errors.Trace(err)
-	}
-	cons := StorageConstraints{
-		Pool:  params.Pool,
-		Size:  params.Size,
-		Count: 1,
+func (st *State) volumeParamsWithDefaults(params VolumeParams, machineId string) (VolumeParams, error) {
+	if params.Pool == "" {
+		envConfig, err := st.ModelConfig()
```
Comment (wallyworld) [inline state/volume.go]: nit: modelConfig

File: state/volume_test.go [hunk 12]
```diff
@@ -563,6 +573,19 @@ func (s *VolumeStateSuite) TestDestroyVolume(c *gc.C) {
...
+func (s *VolumeStateSuite) TestDestroyVolumeStorageAssigned(c *gc.C) {
+	volume, _ := s.setupStorageVolumeAttachment(c)
+	storageTag, err := volume.StorageInstance()
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.State.DestroyVolume(volume.VolumeTag())
+	c.Assert(err, gc.ErrorMatches, "destroying volume 0/0: volume is assigned to storage data/0")
+
+	removeStorageInstance(c, s.State, storageTag)
+	err = s.State.DestroyVolume(volume.VolumeTag())
```
Comment (wallyworld) [inline state/volume_test.go]: replace these 2 lines by assertDestroy()

File: state/cleanup.go (hunk 3 above)
Comment (axw) [inline state/cleanup.go]: I don't follow. DestroyVolume returns a typed error if the volume contains a filesystem. If you destroy the filesystem, and the volume is bound to the filesystem, the volume will be destroyed when the filesystem is removed from the model.

File: state/cleanup.go (hunk 1 above)
Comment (axw) [inline state/cleanup.go]: Done

File: state/filesystem.go (hunk 2 above)
Comment (axw) [inline state/filesystem.go]: A filesystem is not necessarily attached to a machine, and is never directly associated with a unit. The associated storage could potentially be attached to multiple units. I'm going to leave this for now, unless you have strong feelings about it.

File: state/filesystem.go (hunk 4 above)
Comment (axw) [inline state/filesystem.go]: Fixed.

File: state/filesystem_test.go (hunk 6 above)
Comment (axw) [inline state/filesystem_test.go]: I don't understand. There is no assertDestroy function.

File: state/model.go (hunk 7 above)
Comment (axw) [inline state/model.go]: They mean different things. A volume "in" the model means that it is tracked by the model. If it is "bound to" the model, it means that it dies when the model dies.

File: state/model_test.go (hunk 8 above)
Comment (axw) [inline state/model_test.go]: Fixed.

File: state/model_test.go (hunk 9 above)
Comment (axw) [inline state/model_test.go]: The only way you can have non-persistent storage is by having it bound to some non-model entity. There are lifecycle binding tests elsewhere.

Comment (axw) [inline state/model_test.go]: (Also: done, added the comment)

File: state/volume.go (hunk 11 above)
Comment (axw) [inline state/volume.go]: done

File: state/cleanup.go (hunk 3 above)
Comment (wallyworld) [inline state/cleanup.go]: Sorry, I misread errors.Trace() and errors.Errorf() for some reason.

File: state/filesystem.go (hunk 2 above)
Comment (wallyworld) [inline state/filesystem.go]: Let's release as is and get feedback. Maybe just add an extra line to the error message telling the user to juju show-storage to get more details on what in the model is using the storage. An average user will have no idea what "assigned to <id>" means.

File: state/filesystem_test.go (hunk 6 above)
Comment (wallyworld) [inline state/filesystem_test.go]: Ah, hard to tell from the diff. There's a local func defined on line 526. This destroys and checks life. Perhaps that func could be extracted to live on the test struct and used in multiple tests.

File: state/model.go (hunk 7 above)
Comment (wallyworld) [inline state/model.go]: Ok, subtle.

File: state/filesystem_test.go (hunk 6 above)
Comment (axw) [inline state/filesystem_test.go]: I see. That one exists for testing concurrency. Probably not worth extracting, but I'll look at doing the same thing here.

Comment (axw) [inline state/filesystem_test.go]: Hmm actually no, doesn't really make sense here, since you can't yet assign a filesystem to a storage instance post-creation. If you were able to do that, we could do a concurrent DestroyFilesystem and assign-to-storage.

I've extracted the DestroyFilesystem & life check, and used it throughout the suite.

**Comment by axw** [comment]
$$merge$$
